packages/fleather/lib/src/widgets/history.dart

```diff
@@ -0,0 +1,280 @@
+import 'dart:async';
+
+import 'package:fleather/fleather.dart';
+import 'package:flutter/widgets.dart';
+import 'package:quill_delta/quill_delta.dart';
+
+/// A void function that takes a [TextEditingValue].
+@visibleForTesting
+typedef TextEditingValueCallback = void Function(TextEditingValue value);
+
+/// Provides undo/redo capabilities for text editing.
+///
+/// Listens to [controller] as a [ValueNotifier] and saves relevant values for
+/// undoing/redoing. The cadence at which values are saved is a best
+/// approximation of the native behaviors of a hardware keyboard on Flutter's
+/// desktop platforms, as there are subtle differences between each of these
+/// platforms.
+///
+/// Listens to keyboard undo/redo shortcuts and send update to [controller].
+class FleatherHistory extends StatefulWidget {
+  /// Creates an instance of [FleatherHistory].
+  const FleatherHistory(
+      {super.key, required this.child, required this.controller});
+
+  /// The child widget of [FleatherHistory].
+  final Widget child;
+
+  /// The [FleatherController] to save the state of over time.
+  final FleatherController controller;
+
+  @override
+  State<FleatherHistory> createState() => _FleatherHistoryState();
+}
+
+class _FleatherHistoryState extends State<FleatherHistory> {
+  late final _Throttled<Delta> _throttledPush;
+  late HistoryStack _stack;
+  Timer? _throttleTimer;
+
+  // This duration was chosen as a best fit for the behavior of Mac, Linux,
+  // and Windows undo/redo state save durations, but it is not perfect for any
+  // of them.
+  static const Duration _kThrottleDuration = Duration(milliseconds: 500);
+
+  void _undo(UndoTextIntent intent) {
+    _update(_stack.undo());
+  }
+
+  void _redo(RedoTextIntent intent) {
+    _update(_stack.redo());
+  }
+
+  void _update(Delta? changeDelta) {
+    if (changeDelta == null || changeDelta.isEmpty) {
+      return;
+    }
+
+    widget.controller.compose(changeDelta,
+        selection: _fromDelta(changeDelta), source: ChangeSource.local);
+  }
+
+  TextSelection _fromDelta(Delta changeDelta) {
+    assert(changeDelta.isNotEmpty);
+    final firstOp = changeDelta.first;
+    int baseOffset = 0;
+    if (firstOp.isRetain && firstOp.attributes == null) {
+      baseOffset = firstOp.length;
+    }
+    int extentOffset = baseOffset;
+    final lastOp = changeDelta.last;
+    if (lastOp.isRetain && lastOp.attributes != null) {
+      extentOffset = changeDelta.textLength;
+    }
+    if (lastOp.isInsert) {
+      baseOffset = changeDelta.textLength;
+      extentOffset = baseOffset;
+    }
+    return TextSelection(baseOffset: baseOffset, extentOffset: extentOffset);
+  }
+
+  void _push() {
+    if (widget.controller.plainTextEditingValue == TextEditingValue.empty) {
+      return;
+    }
+    _throttleTimer = _throttledPush(widget.controller.document.toDelta());
```

Comment:
why not using changes instead of diffing full documents?

(and aggregate them to the last element of the stack if needed)

Reply:
would love to 
Any idea where to source those changes?
I was thinking we could enhance controller once we move to leveraging on `DeltaTextEditingValue`

Reply:
```controller.document.changes```

Reply:
Ok but then you will capture all changes even those coming from a remote source

Reply:
I'll give it a try (sorry this comes after a long (much interrupted) thought)

Reply:
remote or not you need to capture them, you probably only want to undo your local change, but you will need to rebase to transpose them over the remote changes.

Reply:
Yes thought so too

Reply:
I'm starting to implement but several questions arise:

1. Rebasing is not that simple. I enter `hello world` then I format in **bold** .
   Someone overwrites all with `bye`
   
   **What is the expected behavior?**
2. **If** we want to mimic the thottling that Flutter has implemented (which I believe we should but shout if you think otherwise), we need to implement a coalescing mechanism whereby when we compose any sequence of local `ParchmentChange` and any sequence of remote `ParchmentChange` during the throttle timer
   (`{loc, loc loc},{remote,remote},{loc}...`) to treat them at the end of the timer.

   While this is possible, it sounds overcomplicated for a first version

Happy to read your thoughts on this

Reply:
1. Ok but we need to agree on a algo to handle all cases
2. Not complicated but given the above I think we should leave it aside and handle remote changes in a separate PR
3. I think we should review the source of changes in general not only for the undo/redo, again a separate PR. We might face performance issues on very long documents so I'd adress this first

Reply:
1. I believe this is already coded properly. (just use Delta.transform)
2. yep
3. how would you implement it without knowing the change is coming from undo? adding some kind of metadata to the change? comparing md5?

Reply:
1. Yes `transform` is convenient, but we need to decide who takes precedence and when etc... more a specification work

3. Dunno yet. Maybe simply add a `undo_redo` to `hangeSource`. Didn't think this through